_posts/2017/2017-06-02-reactive-streams.md

```diff
@@ -0,0 +1,149 @@
+But as a collection can be picked up as a whole, a stream has a temporal component.
+Even though one might have a stream at hand, the individual elements in that stream might not yet be available.
+And other elements might already have passed away.
+A reactive stream is also just a stream but with another addition, the so-called "back-pressure".
+Back-pressure means, that the upstream part of a stream only produces data when the down-stream part signaled it's readiness.
+This concept is really not new.
+A classical example for a back-pressured stream is a simple file download.
+
+{% imagebasic blog/blog-reactive-streams-download-example.png Filedownload as reactive stream %} image {% endimagebasic %}
+
+The webserver will send you one chunk of bytes after the other.
+But at any given time the server will only have loaded as much data from disk into memory as you as a client are capable of retrieving.
+How fast you can download the data depends on the speed of your internet (which will most likely be the slowest link in the chain).
+But if you have 10GB network, then it might be faster and your webbrowser might be the bottleneck due to CPU limitations.
+And if the CPU is also fast enought then your good old magnetic harddrive will be the bottleneck.
+
+Let's suppose the harddrive is the bottleneck.
+If every other stage in the chain keeps running as fast as it can, then data will pile up in your webbrowser.
+Eventually your memory is filled up and the browser only has to bad options: Drop byte chunks or completely cancel the download.
+I guess you never had a download cancelled, because the webserver was too fast.
+Maybe because you never had internet that is faster than your disk drive.
+But even if: Back-pressure prevents this kind of problems.
+No matter which is the slowest stage in a reactive streams chain, the other stages will slow down accordingly, so that every individual stage can catch up and does not have to cache extranoues data elements which have already flowed in, but cannot yet be processed.
+
+The back-pressure is achived by a simple twist: Every stage keeps a demand counter that is initially 0.
+Only if the downstream demand is positive it is allowed to send elements downstream.
+To stay in our picture of the download: The first thing that happens, is that the client's disks signals demand, saying it is ready to store data.
+This demand is propagated upstream to the webbrowser, from there over the internet and so on.
+When the demand hits the server's disk, data is read and sent downstream to the webserver, from there over the internet and so on.
+If one stage is slowing down, the whole downstream-data/upstream-demand cycle slows down.
+
+## Coding example with akka-streams
+
+The reactive-streams API is not intended for direct usage, but as interop-layer between individual frameworks.
+A non-exhaustive unordered list is
+
+* [Akka streams][akka-streams],
+* [RxJava][rx-java],
+* [Vert.x][vertx]
```

Comment:
If not a Java-specific list you missed brand new [GenStage](https://github.com/elixir-lang/gen_stage) 😺